File: Leetcode/Python/23. Merge k Sorted Lists/solution.py
# https://leetcode.com/problems/merge-k-sorted-lists/


# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
import heapq


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        heap = []
        root = cur = ListNode(0)
        for idx in range(len(lists)):
            if lists[idx]:  # Filter blank linked list
                # ListNode cant compare so add index between value and listNode
                heapq.heappush(heap, (lists[idx].val, idx, lists[idx]))
        while heap:
            node = heapq.heappop(heap)  # (value, index, linked list)
            idx = node[1]
            cur.next = node[2]
            cur = cur.next

            if cur.next:
                heapq.heappush(heap, (cur.next.val, idx, cur.next))
        return root.next

File: Leetcode/Java/23. Merge k Sorted Lists/solution.java
// https://leetcode.com/problems/merge-k-sorted-lists/submissions/1296204625/

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    // After inserting the nodes into the priority queue, retrieve the node with the minimum value and connect it to the root node
    public ListNode mergeKLists(ListNode[] lists) {
        PriorityQueue<ListNode> priorityQ = new PriorityQueue<>((n1, n2) -> {
            if (n1.val == n2.val){
                return 0;
            } else if (n1.val > n2.val){
                return 1;
            } else {
                return -1;
            }
        });

        for (ListNode node : lists){
            if (node != null){
                priorityQ.add(node);
            }
        }

        ListNode root = new ListNode(-1);
        ListNode cur = root;

        while (!priorityQ.isEmpty()){
            ListNode node = priorityQ.poll();
            cur.next = node;
            cur = cur.next;

            if (node.next != null){
                priorityQ.add(node.next);
            }
        }

        return root.next;


    }
}